happi/audit.py

```diff
@@ -0,0 +1,140 @@
+"""
+This module contains functions and helpers for auditing the happi database.
+Checks are simple functions that take a happi.SearchResult a positional
+argument and returns None when successful.  When a check fails, it should throw
+an Exception with a helpful error message.  These exception messages will be
+caught and organized by the cli audit tool.
+"""
+import inspect
+from typing import Callable, List, Optional, Tuple
+
+from happi import SearchResult
+
+CLIENT_KEYS = ['_id', 'type', 'creation', 'last_edit']
+
+
+def check_instantiation(result: SearchResult) -> None:
+    """
+    Check if the device can be instantiated
+    Simply attempts to get the device from the happi.SearchResult
+
+    Subscription callback exceptions are silenced in ophyd, and will
+    not cause this check to fail.
+    """
+    result.get()
+
+
+def check_extra_info(
+    result: SearchResult,
+    ignore_keys: Optional[List[str]] = None
+) -> None:
+    """
+    Check if there is any extra info in the result that does not
+    have a corresponding EntryInfo on the container.
+
+    Ignores the presence of client-side metadata keys
+    - id
+    - type
+    - creation
+    - last_edit
+    """
+    if ignore_keys is None:
+        ignore_keys = CLIENT_KEYS
+
+    extra = result.item.extraneous.copy()
+    for key in ignore_keys:
+        extra.pop(key)
+    if len(extra) != 0:
+        raise ValueError(f'Un-enforced metadata found: {list(extra.keys())}')
+
+
+def check_name_match_id(result: SearchResult) -> None:
+    """
+    Check if the item's ``_id`` field matches its ``name``.
+    This is a convention we have held for a while, making searching
+    in either direction easier.
+    """
+    id_field = result.metadata.get('_id')
+    name_field = result.metadata.get('name')
+    assert id_field == name_field, f'id: {id_field} != name: {name_field}'
+
+
+def verify_result(
+    result: SearchResult,
+    checks: List[Callable[[SearchResult], None]]
+) -> Tuple[bool, str, str]:
+    """
+    Validate device against the provided checks
+
+    Parameters
+    ----------
+    result : happi.SearchResult
+        result to be verified
+
+    checks : List[Callable[[SearchResult], None]]
+        a list of verification functions that raise exceptions if
+        verification fails.
+
+    Returns
+    -------
+    bool
+        whether or not the device passed the checks
+    str
+        name of check that failed, if any
+    str
+        error message describing reason for failure and possible steps
+        to fix.  Empty string if validation is successful
+    """
+    success, check, msg = True, "", ""
+    # verify checks have the correct signature, as much as is reasonable
+    for check in checks:
+        validate_check_signature(check)
+
+    try:
+        for check in checks:
+            check(result)
+    except Exception as ex:
```

Comment:
Would there be a benefit to showing _all_ checks that failed instead of just the first check?

Reply:
When I started this I think I was envisioning subsequent checks would be nonsensical if earlier checks failed (checking if methods exist on devices only makes sense if the device can be instantiated).  But with the given checks, I think it makes sense to run all the checks and show all failures.

Reply:
I think what you have proposed is plenty good for a first pass. This could be a follow-up or just something that we file away in an issue - your call

Reply:
addressed in [c2a4bc9](https://github.com/pcdshub/happi/pull/286/commits/c2a4bc95c34499b12ac90d45740425c731927ec0)